validator/src/bootstrap.rs

```diff
@@ -616,629 +500,297 @@ mod without_incremental_snapshots {
+        if peer_snapshot_hashes.is_empty() {
+            match newer_cluster_snapshot_timeout {
+                None => newer_cluster_snapshot_timeout = Some(Instant::now()),
+                Some(newer_cluster_snapshot_timeout) => {
+                    if newer_cluster_snapshot_timeout.elapsed().as_secs() > 180 {
+                        warn!("Giving up, did not get newer snapshots from the cluster.");
+                        return None;
                     }
                 }
-                eligible_rpc_peers
-            };
-
-            if !eligible_rpc_peers.is_empty() {
-                let contact_info =
-                    &eligible_rpc_peers[thread_rng().gen_range(0, eligible_rpc_peers.len())];
-                return Some((contact_info.clone(), highest_snapshot_hash));
-            } else {
-                retry_reason = Some("No snapshots available".to_owned());
             }
+            retry_reason = Some("No snapshots available".to_owned());
+            continue;
+        } else {
+            let rpc_peers = peer_snapshot_hashes
+                .iter()
+                .map(|peer_snapshot_hash| peer_snapshot_hash.rpc_contact_info.id)
+                .collect::<Vec<_>>();
+            let PeerSnapshotHash {
+                rpc_contact_info: final_rpc_contact_info,
+                snapshot_hash: final_snapshot_hash,
+            } = get_final_peer_snapshot_hash(&peer_snapshot_hashes);
+            info!(
+                "Highest available snapshot slot is {}, available from {} node{}: {:?}",
+                final_snapshot_hash
+                    .incr
+                    .map(|(slot, _hash)| slot)
+                    .unwrap_or(final_snapshot_hash.full.0),
+                rpc_peers.len(),
+                if rpc_peers.len() > 1 { "s" } else { "" },
+                rpc_peers,
+            );
+
+            return Some(GetRpcNodeResult {
+                rpc_contact_info: final_rpc_contact_info,
+                snapshot_hash: Some(final_snapshot_hash),
+            });
         }
     }
+}
 
-    fn get_known_snapshot_hashes(
-        cluster_info: &ClusterInfo,
-        known_validators: &Option<HashSet<Pubkey>>,
-    ) -> Option<HashSet<(Slot, Hash)>> {
-        if let Some(known_validators) = known_validators {
-            let mut known_snapshot_hashes = HashSet::new();
-            for known_validator in known_validators {
-                cluster_info.get_snapshot_hash_for_node(known_validator, |snapshot_hashes| {
-                    for snapshot_hash in snapshot_hashes {
-                        known_snapshot_hashes.insert(*snapshot_hash);
-                    }
-                });
+/// Get the Slot and Hash of the local snapshot with the highest slot.  Can be either a full
+/// snapshot or an incremental snapshot.
+fn get_highest_local_snapshot_hash(
+    full_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_fetch: bool,
```

Comment:
I don't think this function needs to change. If a node already has previous incremental snapshots, I don't see a reason _not_ to use them here. The new flag is whether or not to _download new_ incremental snapshots.

Reply:
Yes, but now if I start validator with ‘--no-incremental-snapshots’ than it ignores any incremental snapshot and calculates the snapshot age only by full snapshots (and if it’s not ok, then download new fresh full one). And as this PR is just only about refactoring I would like to save functionality as is.

Reply:
That's fair. I think your impl is good. If we decide a different behavior is desirable, we can change it later.